File: l1/z1/candidate.py
import random

class Candidate:
    def __init__(self, minimum, maximum):
        self.minimum = minimum
        self.maximum = maximum

    def gen(self, size):
        random.seed()
        return [random.uniform(self.minimum, self.maximum) for _ in range(size)]

    def tweak(self, point):
        pass

    def analyze(self, point, newPoint, f):
        pass

class GaussianSteepestAscent(Candidate):
    def tweak(self, point, variance):
        prob = 1.0
        newPoint = [x for x in point]

        for i, x in enumerate(newPoint):
            if random.random() < prob:
                n = random.gauss(0, variance)
                while not (self.minimum <= x + n <= self.maximum):
                    n = random.gauss(0, variance)
                newPoint[i] = x + n

        return newPoint

    def analyze(self, point, newPoint, f, variance):
        w = self.tweak(point, variance)
        if f(w) < f(newPoint):
            return w
        else:
            return newPoint
                

File: l1/z1/cpp/candidate.cpp
#include "candidate.h"
#include <stdlib.h>
#include <time.h>  

using namespace std;

double* Candidate::gen(double min, double max, int size) {
    srand(time(NULL));
    double point[size];
    const int RANGE = 1000000;

    for (int i = 0; i < size; i++) {
        double x = (rand() % RANGE) * (max - min) / RANGE + min;
        point[i] = x;
    }
    return point;
}